File: backend/src/FinanceTracker.Application/Common/Permissions.cs
using FinanceTracker.Domain;

namespace FinanceTracker.Application.Common;

public static class Permissions
{
    public const string TransactionCreate = "transaction.create";
    public const string TransactionApprove = "transaction.approve";
    public const string TransactionDelete = "transaction.delete";
    public const string BudgetManage = "budget.manage";
    public const string ReportView = "report.view";
    public const string MemberInvite = "member.invite";
    public const string MemberManage = "member.manage";
    public const string DepartmentManage = "department.manage";
    public const string AuditView = "audit.view";
    public const string SettingsManage = "settings.manage";
    public const string OrganisationModeSwitch = "organisation.mode.switch";
}

public interface IPermissionService
{
    bool Can(MemberRole role, string permission);
}

public class PermissionService : IPermissionService
{
    private static readonly Dictionary<string, HashSet<MemberRole>> _matrix = new()
    {
        [Permissions.TransactionCreate]      = new() { MemberRole.Owner, MemberRole.Admin, MemberRole.Manager, MemberRole.Member },
        [Permissions.TransactionApprove]     = new() { MemberRole.Owner, MemberRole.Admin, MemberRole.Manager },
        [Permissions.TransactionDelete]      = new() { MemberRole.Owner, MemberRole.Admin, MemberRole.Member },
        [Permissions.BudgetManage]           = new() { MemberRole.Owner, MemberRole.Admin, MemberRole.Manager },
        [Permissions.ReportView]             = new() { MemberRole.Owner, MemberRole.Admin, MemberRole.Manager, MemberRole.Member, MemberRole.Viewer },
        [Permissions.MemberInvite]           = new() { MemberRole.Owner, MemberRole.Admin },
        [Permissions.MemberManage]           = new() { MemberRole.Owner, MemberRole.Admin },
        [Permissions.DepartmentManage]       = new() { MemberRole.Owner, MemberRole.Admin },
        [Permissions.AuditView]              = new() { MemberRole.Owner, MemberRole.Admin },
        [Permissions.SettingsManage]         = new() { MemberRole.Owner, MemberRole.Admin },
        [Permissions.OrganisationModeSwitch] = new() { MemberRole.Owner },
    };

    public bool Can(MemberRole role, string permission)
        => _matrix.TryGetValue(permission, out var allowed) && allowed.Contains(role);
}

File: frontend/src/app/core/auth/permissions.ts
import { MemberRole } from '../models/api.models';

export const Permissions = {
  TransactionCreate: 'transaction.create',
  TransactionApprove: 'transaction.approve',
  TransactionDelete: 'transaction.delete',
  BudgetManage: 'budget.manage',
  ReportView: 'report.view',
  MemberInvite: 'member.invite',
  MemberManage: 'member.manage',
  DepartmentManage: 'department.manage',
  AuditView: 'audit.view',
  SettingsManage: 'settings.manage',
  OrganisationModeSwitch: 'organisation.mode.switch',
} as const;

export type Permission = typeof Permissions[keyof typeof Permissions];

const matrix: Record<string, MemberRole[]> = {
  [Permissions.TransactionCreate]:      ['Owner', 'Admin', 'Manager', 'Member'],
  [Permissions.TransactionApprove]:     ['Owner', 'Admin', 'Manager'],
  [Permissions.TransactionDelete]:      ['Owner', 'Admin', 'Member'],
  [Permissions.BudgetManage]:           ['Owner', 'Admin', 'Manager'],
  [Permissions.ReportView]:             ['Owner', 'Admin', 'Manager', 'Member', 'Viewer'],
  [Permissions.MemberInvite]:           ['Owner', 'Admin'],
  [Permissions.MemberManage]:           ['Owner', 'Admin'],
  [Permissions.DepartmentManage]:       ['Owner', 'Admin'],
  [Permissions.AuditView]:              ['Owner', 'Admin'],
  [Permissions.SettingsManage]:         ['Owner', 'Admin'],
  [Permissions.OrganisationModeSwitch]: ['Owner'],
};

export function canRole(role: MemberRole | null | undefined, permission: string): boolean {
  if (!role) return false;
  return matrix[permission]?.includes(role) ?? false;
}
